Upgrade to react 0.14.0-rc1 and react-router 1.0.0-rc1

stevoland opened the pull request:
An attempt at https://github.com/erikras/react-redux-universal-hot-example/issues/195 and https://github.com/erikras/react-redux-universal-hot-example/issues/182

It's a total hack really, based on using react-router's `RoutingContext` and `match` functions on the client where, apparently, they're not intended to be used. This is the only way to implement the required transition hooks with the new react-router API. Personally I think it would be good to look at using something like https://github.com/ericclemmons/react-resolver. This would mean we wouldn't have to pause the render until the data has loaded so could use the proper react-router API. 

There's bound to be some edge cases I'm not aware of. I also get a weird flicker sometimes when transitioning between routes in dev mode. Anyways, it's a start.

Thanks to @halt-hammerzeit, @quicksnap and @geekyme - I copied their work, basically.

quicksnap commented:
Hey, great work! This is very similar to my fork where I was attempting this as well. 

Interesting approach re-rendering the component tree in tandem with the `preload` option to universalRouter.

In my attempt, I was also using `RoutingContext`, but was attempting to avoid `match` on client side.

quicksnap commented on src/client.js:
```diff
@@ -3,36 +3,43 @@
+};
+
+history.listen((loc) => {
+  render(loc, history, store, true);
```
Is this synchronous? Will the history still `popState` if the render preload fails or hangs?

quicksnap replied:
I just pulled it down and tested it--it seems to be updating the location state before the request completes, which I believe is not how it should operate.

I was looking into Lifecycle hooks to prevent transition so that we could accomplish that: https://github.com/rackt/react-router/issues/2061

stevoland replied:
You're right, I've force pushed a change to use `registerTransitionHook` instead, thanks.

quicksnap commented:
Also, odd bug: I can no longer hide the redux dev tools with Ctrl+H..

stevoland commented:
Hmm, no idea about devtools. Ctrl+H doesn't work for me on master

halt-hammerzeit commented:
@stevoland hmmm, I see, so you chose to render `<RoutingContext {...renderProps}/>` both on server and client (as opposed to `<Router>` on client).
That makes history transition hooks work.
Interesting idea.

stevoland commented:
@halt-hammerzeit I can't really see how else to get a history transition hook and the new component tree before the render for the new location?

halt-hammerzeit commented:
@stevoland yeah, me neither

stevoland commented:
redux-router uses `RoutingContext` https://github.com/rackt/redux-router/blob/master/src/ReduxRouter.js#L87 and that's an 'approved' Rackt lib now so maybe there's hope that this technique isn't totally crazy.

I see this PR as a temporary measure though. I don't think the UI should be blocked until data is loaded. I think the transition should happen and the new containers should be able to display spinners or whatever. I suppose ultimately Relay will be brought in to take care of all this.

halt-hammerzeit commented:
@stevoland yeah, i'm thinking more about Relay too, and that's the reason i'm not working that much on this issue: when i'm done studying relay i'll swap API with it, and that can potentially change a lot of things. I currently haven't yet researched things like "Relay isomorphic rendering" and "Relay + React-Router boiler plate" stuff.

erikras commented:
Relay doesn't feel too compatible with Redux...or Flux at all, really.

stevoland commented:
No, but I think it's on the roadmap to incorporate managing local state into Relay so it will replace Flux in complex apps I think. It's server rendering story is excellent, once it's released: https://github.com/facebook/relay/issues/36. Anyway we're a bit off topic I think!

quicksnap commented:
I personally don't think the UI should block, but one should be able to block a route transition if their app calls for it.

For instance:

* Click on a link, see a "Loading" spinner.
* Get data back from server
* Then, transition to `/widgets` route

Seems like a totally reasonable UX

quicksnap commented:
To that end, we would still need the same transition hook mechanisms.

nicolabortignon commented:
@quicksnap devtool stopped working with chrome 47, there is a bug in the way keys combo are detected

justingreenberg commented:
@stevoland nice work! quick note.. arguments passed to `onEnter` transition hook were updated. new shape is `onEnter(nextState, replaceState, callback)`, i think this only affects the one static method on `containers/RequireLogin` component, ie...

```javascript
  static onEnter(store) {
    return (nextState, replaceState) => {

      const { auth: { user }} = store.getState();
      if (!user) {
        // oops, not logged in, so can't be here!
        replaceState(null, '/');
      }
    };
  }
```

bertramtruong commented:
@stevoland Thanks so much for this! I'm taking a look now at this branch against my work and it seems to be working fine.

Like @justingreenberg mentioned, the current branch will give a warning:

`Warning: registerTransitionHook is deprecated; use listenBefore instead`

I replaced the line in `client.js` from:

    history.registerTransitionHook((loc, callback) => {
        render(loc, history, store, true)
            .then((callback));
    });

to:

    history.listenBefore((loc, callback) => {
        render(loc, history, store, true).then(callback);
    });

and the warning disappeared :)

stevoland commented:
@justingreenberg @bertramtruong Thanks, I've updated the branch.

Fixed another couple of warnings and rebased with latest master.

@erikras What do you think?

erikras commented:
Excellent work, you guys. @stevoland is the man!

Looks good enough to merge to me.

quicksnap commented:
:confetti_ball: :tada: :balloon: :dancers: :dancer: 

I'm going to be using this as my base today; excited to see how it pans out!

Great work everyone! Thanks @stevoland!

erikras commented:
Seriously. @stevoland deserves a :beer: or two.

geekyme commented:
Hey I think this part https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/client.js#L56 doesn't seem to be be right. How come the entire tree is rerendered on route transition. Also, `createRoutes` will be called every single time someone navigates. Seems expensive.

stevoland commented:
There's some room for improvement for sure, maybe a wrapper component that
listens to history. I might do this if I look into integrating redux-router.

Performance wise I wouldn't worry.
On 26 Sep 2015 03:59, "Shawn Lim" <notifications@github.com> wrote:

> Hey I think this part
> https://github.com/erikras/react-redux-universal-hot-example/blob/master/src/client.js#L56
> doesn't seem to be be right. How come the entire tree is rerendered on
> route transition. Also, createRoutes will be called every single time
> someone navigates. Seems expensive.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/erikras/react-redux-universal-hot-example/pull/214#issuecomment-143393447>
> .
>

apapirovski commented:
FWIW, I've got a version that runs as a middleware component between the router and the App itself. You can check it out at https://github.com/rackt/react-router/issues/2101

It would need to be updated if you're not doing server-side rendering because it completely bypasses the initial mounting on client-side but it has worked quite well for me.

stevoland commented:
@apapirovski Very nice!
